### ISSUE TITLE
class of element in data_frame is dropped when subsetting with `[`

### ISSUE THREAD

**Issue by tinu-schneider** [hadley/dplyr#2172]
## tl;dr

```
dat <- c("a", "b")

d.f <-        data.frame('dat' = dat, stringsAsFactors = FALSE)
d_f <- dplyr::data_frame('dat' = dat)

# Fails: "Error: d.f[, 1] not equal to d_f[, 1]"
testthat::expect_equal(d.f[, 1], d_f[, 1])

# Passes
testthat::expect_equal(d.f$dat, d_f$dat)

# Problem with dropped class, when subsetting with '['
# class(d.f[, 1]): "character"
# class(d_f[, 1]): "tbl_df" "tbl" "data.frame"
```
## Long Version

I want to test the class of a column in a `data_frame`. If I subset with `[` the class of the column is dropped. So I can't test for class `character` (which I need to).

It seems as if this is related to [tibble #155](https://github.com/tidyverse/tibble/issues/155) and [dplyr #1984](https://github.com/hadley/dplyr/issues/1984).

To show this, I create a base `data.frame` and a dplyr `data_frame` with one column containing the two values `a` and `b`. If I subset the column with `[, 1]`, the classes are not the same. **But** if I subset with `$names` the classes are equal.

```
# create data
std <- data.frame('names' = c('a', 'b'), stringsAsFactors = FALSE)
dpl <- data_frame('names' = c('a', 'b'))
```

If I subset with `std[, 1], dpl[, 1]` I get

```
Elements:
[1] "a" "b"
# A tibble: 2 × 1
  names
  <chr>
1     a
2     b

Class of elements:
[1] "character"
[1] "tbl_df"     "tbl"        "data.frame"
```

But if I subset with `std$names, dpl$names` the result is as expected

```
Elements:
[1] "a" "b"
[1] "a" "b"

Class of elements:
[1] "character"
[1] "character"
```

Here is the rather lengthy code to reproduce the example.

```
library(dplyr)
library(testthat)

# helper
helper <- function(x1, x2) {
    cat('\nElements:\n')
    print(x1)
    print(x2)
    cat('\nClass of elements:\n')
    print(class(x1))
    print(class(x2))
}

# create data
std <- data.frame('names' = c('a', 'b'), stringsAsFactors = FALSE)
dpl <- data_frame('names' = c('a', 'b'))

cat('\nA: Subset with "["')
x1 <- std[, 1]
x2 <- dpl[, 1]
helper(x1, x2)

cat('\n\nB: Subset with "$"')
y1 <- std$names
y2 <- dpl$names
helper(y1, y2)

cat('\n\nC: Subset with "$" AND "["')
z1 <- std$names[1]
z2 <- dpl$names[1]
helper(z1, z2)

# And some tests
# '[' fails
expect_equal(x1, x2)
expect_equal(class(x1), class(x2))

# '$' works
expect_equal(y1, y2)
expect_equal(class(y1), class(y2))

# '$' and '[' works
expect_equal(z1, z2)
expect_equal(class(z1), class(z2))
```

**Comment by krlmlr** [comment]
Unlike data frames, tibbles never drop dimensions. This is by design. Calling `[, 1]` returns a vector on a data frame, and a one-column tibble on a tibble.

```r
testthat::expect_equal(d.f[, 1, drop = FALSE], as.data.frame(d_f[, 1]))
```

**Comment by tinu-schneider** [comment]
Thanks Kirill, this helps. I wasn't aware, that 'tibbles' work in the background of dplyr, but after reading the tibble vignettes it is now clear. 

So, if want to check for the class of the column of a tibble, I use `[[` to subset.

    testthat::expect_equal(d.f[, 1], d_f[[1]])

    is.vector(d.f[, 1])
    [1] TRUE
    is.vector(d_f[[1]])
    [1] TRUE

And finally

    class(d_f[[1]])
    [1] "character"

which is, what I was looking for.
